### PR TITLE
Ckan save

### PR DESCRIPTION
**Pull request by sottom**

### CODE REVIEW

File: app/api.py [hunk 1]
```diff
@@ -165,6 +163,64 @@ def get_error_resp(e, status_code=500):
...
+            },
+            "files": [('upload', src)]
+        }
+
+    if not response:
+        # upload data
+        post_data['headers'] = headers
+        r = requests.post(**post_data)
+        resp = r.json()
+        if r.status_code is 200:
```
Comment (reubano) [inline app/api.py:208]: `r.ok`

File: app/api.py [hunk 2]
```diff
@@ -430,11 +491,12 @@ def fetch_report(report_date, enqueue=False, **kwargs):
...
         response = enqueue_work(report_date, **kwargs)
     else:
+        # TODO: change all locations that use 'use_s3' for save_report function to use 'dest' instead
+        # TODO: Also possibly change get_report
         get_from_s3 = kwargs.get("source") == "s3"
-        save_to_s3 = kwargs.get("dest") == "s3"
 
         report = get_report(report_date, use_s3=get_from_s3, **kwargs)
-        json = save_report(report, report_date, use_s3=save_to_s3, **kwargs)
+        json = save_report(report, report_date, save_dest=kwargs.get("dest"), **kwargs)
```
Comment (reubano) [inline app/api.py:499]: try to avoid nested kwargs...

File: dev-requirements.txt [hunk 3]
```diff
@@ -4,7 +4,6 @@ coverage==4.3.4
 flake8==3.7.9
 pytest==4.3.0
 pylint==1.9.1
-pkutils==0.13.6
```
Comment (reubano) [inline dev-requirements.txt:7]: why remove this?

File: app/api.py [hunk 4]
```diff
@@ -572,10 +637,10 @@ def post(self):
             start_date = self.end_date - timedelta(days=day)
             report_date = start_date.strftime(S3_DATE_FORMAT)
             response = fetch_report(report_date, **self.kwargs)
-            json = response.get("result")
+            json = response.pop('result', {})
 
             if json:
-                result[report_date] = json
+                result[report_date] = {**json, **response}
```
Comment (reubano) [inline app/api.py:643]: not sure I understand this change...

File: app/api.py [hunk 5]
```diff
@@ -165,6 +164,67 @@ def get_error_resp(e, status_code=500):
...
+                "name": filename,
+            },
+            "files": [('upload', src)]
+        }
+
+    if not response:
+        # upload data
+        post_data['headers'] = headers
+        r = requests.post(**post_data)
+        resp = r.json()
```
Comment (reubano) [inline app/api.py:211]: should be `json = r.json()`

File: app/api.py [hunk 6]
```diff
@@ -165,6 +164,67 @@ def get_error_resp(e, status_code=500):
...
 
+def save_ckan_report(src, report_date, **kwargs):
+    report_type = kwargs["report_type"]
+    config = REPORT_CONFIGS[report_type]
+    filename = config["filename"].format(report_date)
+    headers = {"X-CKAN-API-Key": CKAN_API_KEY}
+    response = None
+
+    # check for data before uploading it
+    resp = requests.post(
```
Comment (reubano) [inline app/api.py:175]: always set the response of `requests` to `r`.
```
r = requests.post()
json = r.json()
```

File: app/api.py [hunk 7]
```diff
@@ -165,6 +164,67 @@ def get_error_resp(e, status_code=500):
...
+        f"{CKAN_API_BASE_URL}/resource_search",
+        data={"query": f"name:{filename}"},
+        headers=headers,
+    ).json()
+
+    action = 'updated' if resp['result']['count'] else 'created'
+
+    if not kwargs.get('overwrite') and action == 'updated':
+        response = get_error_resp(f"{filename} already exists!")
+    elif kwargs.get('overwrite') and action == 'updated':
```
Comment (reubano) [inline app/api.py:187]:
```python
overwrite = kwargs.get('overwrite')
is_update = action == 'updated'

if overwrite and is_update:
    pass
elif is_update:
    pass
else:
    pass
```

File: app/api.py (hunk 4 above)
Comment (sottom) [reply app/api.py:643]: Output with this change:

```json
{
  "message": "IL_county_COVID19_data_20200616.json already exists!",
  "ok": false,
  "result": {
    "20200616": {
      "content_length": null,
      "etag": null,
      "last_modified": null,
      "message": "IL_county_COVID19_data_20200616.json already exists!",
      "ok": false,
      "status_code": 500
    },
    "20200617": {
      "content_length": null,
      "etag": null,
      "last_modified": null,
      "message": "IL_county_COVID19_data_20200617.json already exists!",
      "ok": false,
      "status_code": 500
    },
    "20200618": {
      "content_length": null,
      "etag": "IL_county_COVID19_data_20200618.json",
      "last_modified": "2020-06-22T16:38:22.782249",
      "message": "Successfully created IL_county_COVID19_data_20200618.json!",
      "ok": true,
      "status_code": 200
    },
    "20200619": {
      "content_length": null,
      "etag": "IL_county_COVID19_data_20200619.json",
      "last_modified": "2020-06-22T16:38:20.006511",
      "message": "Successfully created IL_county_COVID19_data_20200619.json!",
      "ok": true,
      "status_code": 200
    },
    "20200620": {
      "content_length": null,
      "etag": "IL_county_COVID19_data_20200620.json",
      "last_modified": "2020-06-22T16:38:17.394866",
      "message": "Successfully created IL_county_COVID19_data_20200620.json!",
      "ok": true,
      "status_code": 200
    },
    "20200621": {
      "content_length": null,
      "etag": "IL_county_COVID19_data_20200621.json",
      "last_modified": "2020-06-22T16:38:14.741343",
      "message": "Successfully created IL_county_COVID19_data_20200621.json!",
      "ok": true,
      "status_code": 200
    },
    "20200622": {
      "content_length": null,
      "etag": null,
      "last_modified": null,
      "message": "No data found for 20200622.",
      "ok": false,
      "status_code": 404
    }
  },
  "status": "Internal Server Error"
}
```

Comment (sottom) [reply app/api.py:643]: I will add a comment above the pop to explain the reasoning (prevent nesting).

File: app/api.py (hunk 6 above)
Comment (sottom) [reply app/api.py:175]: Will naming the variable `json` become an issue if we `import json`? I guess we could just rename the import.

Comment (reubano) [reply app/api.py:175]: Yes. But I use `from json import...`. :)

Comment (sottom) [reply app/api.py:175]: Ah, I see :) `json` it is then!

**Pull request merged**